src/meta/processors/admin/Balancer.cpp

```diff
@@ -0,0 +1,316 @@
+            auto status = buildBalancePlan();
+            if (plan_ == nullptr) {
+                LOG(ERROR) << "Create balance plan failed!";
+                return status;
+            }
+        }
+        executor_->add(std::bind(&BalancePlan::invoke, plan_.get()));
+        return plan_->id();
+    }
+    return Status::Error("balance running");
+}
+
+bool Balancer::recovery() {
+    CHECK(!plan_) << "plan should be nullptr now";
+    if (kv_) {
+        const auto& prefix = BalancePlan::prefix();
+        std::unique_ptr<kvstore::KVIterator> iter;
+        auto ret = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            LOG(ERROR) << "Can't access kvstore, ret = " << static_cast<int32_t>(ret);
+            return false;
+        }
+        std::vector<int64_t> corruptedPlans;
+        while (iter->valid()) {
+            auto balanceId = BalancePlan::id(iter->key());
+            auto status = BalancePlan::status(iter->val());
+            if (status == BalancePlan::Status::IN_PROGRESS
+                    || status == BalancePlan::Status::FAILED) {
+                corruptedPlans.emplace_back(balanceId);
+            }
+            iter->next();
+        }
+        if (corruptedPlans.empty()) {
+            LOG(INFO) << "No corrupted plan need to recovery!";
+            return true;
+        }
+        CHECK_EQ(1, corruptedPlans.size());
+        plan_ = std::make_unique<BalancePlan>(corruptedPlans[0], kv_, client_.get());
+        plan_->onFinished_ = [this] () {
+            plan_.reset();
+            bool expected = true;
+            CHECK(running_.compare_exchange_strong(expected, false));
+        };
+        if (!plan_->recovery()) {
+            LOG(ERROR) << "Can't recovery plan " << corruptedPlans[0];
+            plan_->onFinished_();
+            return false;
+        }
+    }
+    return true;
+}
+
+Status Balancer::buildBalancePlan() {
+    CHECK(!plan_) << "plan should be nullptr now";
+    std::vector<GraphSpaceID> spaces;
+    {
+        // Get all spaces
+        folly::SharedMutex::ReadHolder rHolder(LockUtils::spaceLock());
+        auto prefix = MetaServiceUtils::spacePrefix();
+        std::unique_ptr<kvstore::KVIterator> iter;
+        auto ret = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            running_ = false;
+            return Status::Error("Can't access kvstore, ret = %d", static_cast<int32_t>(ret));
+        }
+        while (iter->valid()) {
+            auto spaceId = MetaServiceUtils::spaceId(iter->key());
+            spaces.push_back(spaceId);
+            iter->next();
+        }
+    }
+    plan_ = std::make_unique<BalancePlan>(time::TimeUtils::nowInSeconds(), kv_, client_.get());
+    for (auto spaceId : spaces) {
+        auto tasks = genTasks(spaceId);
+        for (auto& task : tasks) {
+            plan_->addTask(std::move(task));
```

Comment:
why not define a function “addTasks(vector<task>&&)” in plan?

Reply:
A HA.  There are two reasons:
1.  AddTask is designed firstly.
2.  AddTask is more convenient in UTs